### ISSUE TITLE
Uploading package file encoded as UTF8 with BOM fails with Grails error

### ISSUE THREAD

**Issue by ostephens** [k-int/KBPlus#717]
If a package upload file is "UTF8 with BOM" encoded the upload fails with a Grails error. Re-saving the same package as UTF8 (no BOM) upload succeeds without a problem. Grails error is as follows:

Error 500: Internal Server Error
URI
/kbplus/upload/reviewSO
Class
java.lang.NullPointerException
Message
Cannot get property 'messages' on null object
Trace
    Line | Method
->>  632 | generateAndValidatePackageIdentifier in com.k_int.kbplus.UploadController

---

|    456 | validate                             in     ''
|     61 | reviewSO . . . . . . . . . . . . . . in     ''
|    195 | doFilter                             in grails.plugin.cache.web.filter.PageFragmentCachingFilter
|     63 | doFilter . . . . . . . . . . . . . . in grails.plugin.cache.web.filter.AbstractFilter
|     65 | doFilter                             in com.k_int.kbplus.filter.ApiauthFilter
|   1110 | runWorker . . . . . . . . . . . . .  in java.util.concurrent.ThreadPoolExecutor
|    603 | run                                  in java.util.concurrent.ThreadPoolExecutor$Worker
^    722 | run . . . . . . . . . . . . . . . .  in java.lang.Thread

**Comment by ianibo** [comment]
I'm not sure that CSV _files_ should contain a bom should they?

**Comment by ostephens** [comment]
The problem seems to be that notepad on Windows adds BOM as default when it saves as UTF8. I've recommended the use of an alternative (notepad++) but possibly worth checking for a BOM even if outcome is simply to reject upload?

**Comment by ianibbo** [comment]
Ok - I've wrapped the upload input stream with a BOMInputStream that can
skip BOM if present. This is one of those deep technical & philosophical
issues that I don't have the energy to debate really, so I've just made the
change as requested. Be aware that there are now 2 [Well, three actually]
different ways of specifying file character set encoding when uploading a
file, this makes the process somewhat non-deterministic.

Ian Ibbotson
Director
Knowledge Integration Ltd
35 Paradise Street, Sheffield. S3 8PZ
T: 0114 273 8271
M: 07968 794 630
W: http://www.k-int.com
